### PR TITLE
tests: adding test cases for interrupt

### PR DESCRIPTION
**Pull request by enjiamai**
Add test cases of interrupt for improving test coverage. Some of the test cases for other platform were already existing, so this submission mainly focus on adding the missing portion.

The new test case "test_isr_static" add following tests for all platform except qemu_xtensa:
1. To check parameters pass to interrupt handler.
2. To declare interrupts statically at build time.
3. To validate multiple ISRs utilizing the same function to process interrupts.
4. To validate kernel provide a dedicated interrupt stack for processing hardware interrupts.

And try to add support dynamic interrupt testing for the x86, x86_64, and native_posix platform in original test_dynamic_isr.  

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

### CODE REVIEW

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (ioannisg) [inline tests/kernel/interrupt/src/interrupt_common.c:64]: This seems to me cross-ARCH; can this be done for the other architectures, or is it done already?

**Review by ioannisg — COMMENTED** [review]
Currently this seems to be x86-specific, though I don't fully-understand why,
so i leave this to the x86 maintainers

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c:64]: Most parts of this case can be found in other test cases, such as configuration of interrupts at build time/runtime, or pass argument to interrupt handler. I saw them in tests/kernel/interrupt, tests/kernel/gen_isr_table, etc.
 Should I implement this test case to cross-ARCH? Any suggestion for this?

**Review by stephanosio — CHANGES_REQUESTED** [review]
Overall, needs better abstraction and cross architecture support.

If not, `interrupt_common.c` should really be renamed to something like `interrupt_x86.c` or, even better, be just moved to the arch tests (`tests/arch/x86`).

**Comment by enjiamai** [comment]
@stephanosio Ok, got your point, will update all of this ASAP, thanks for your advice!

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: OK, already move this to interrupt_util.h, and implement the trigger_irq for x86.

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: Ok, also move this to interrupt_util.h. But not implement for other architecture yet. It takes some time to do this, will update it later.

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: Fix it. Thanks for your reminder!

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: @stephanosio I added this to dynamic_isr.c, for testing X86. And do you think other architecture such as posix, risc-v, nios2, arc,should also be covered too?

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: Yes, I totally agree with you. I change the case name to static_isr.c and moved the dynamic test out of here. It only tests x86 now. But I will try to expend this case to fit other architecture later. 
Actually,  tests/kernel/gen_isr_table has already implemented some different platform for interrupt testing, but not for all. I am a little bit confusing here. Should I just try to add more architecture support by improving tests/kernel/gen_isr_table as possible?

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: Fixed it in another way. Just try to check it's allocation is failed or not.

**Comment by enjiamai** [comment]
@stephanosio @ioannisg  I have updated it, could you please help to review it, thanks!

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: tests/kernel/interrupt/src/static_isr.c
Comment (stephanosio) [inline tests/kernel/interrupt/src/static_isr.c:11]: The main problem with static ISR test is that we cannot "dynamically" (as in, at run-time) find an available IRQ, and the available IRQ numbers may vary across the SoCs for some archs, so we cannot hard-code them in the test (unless we `#ifdef` every available SoC or introduce some kind of automation scripts to handle this).

In that sense, I understand that this test needs to be restricted to certain supported arch/SoCs, but we should not be hard-coding x86 here in the static ISR test; instead, such constructs should be abstracted/defined in the `interrupt_util.h` (e.g. checking whether `TEST_IRQ_0_LINE` and `TEST_IRQ_1_LINE` are defined).

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/dynamic_isr.c:78]: OK, got it. It indeed a little trouble here when I implemented this. Let me think how to make this better.

File: tests/kernel/interrupt/src/static_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/static_isr.c:18]: Yes, it shall be. Fix it.

File: tests/kernel/interrupt/src/static_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/static_isr.c:23]: Just understand. Got it fixed.

**Comment by ioannisg** [comment]
what is the status of this PR @enjiamai @stephanosio ?

**Comment by enjiamai** [comment]
@ioannisg So sorry for missing your commented for almost two weeks. Due to some reason, I am just working back on it these days. 
I considered the suggestion what @stephanosio give, it shouldn't be hard code for static interrupt and make the test cases more common here (arch-independent) is important. But I didn't come up with a better way to do it. So I will just simply use #ifdef to do it.

File: tests/kernel/interrupt/src/static_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/static_isr.c]: I make some change, to put the trigger_irq() function in interrupt_util.h. If there are still things need to be improve, pls kindly let me know, thanks.

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/dynamic_isr.c]: Yes, this "__unused" should not be used here, fixed it.

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: Yes, fixed it.

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: yes, remove it. It's already in dynamc_isr.c.

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/dynamic_isr.c]: fixed it.

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: fixed it. Put some missing case in dynamic_isr.c already.

File: tests/kernel/interrupt/src/interrupt_common.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_common.c]: This file name has been updated. Fixed.

**Review by LeiW000 — APPROVED** [review]

**Comment by enjiamai** [comment]
@stephanosio @ioannisg Sorry for reply late. I have solved all the comments that you offered.
 The most important part is:
1. Adding missing platform that support dynamic ISR. 
2. Static ISR for all platform except xtensa and nios2. For xtensa, I don't know why I can't trigger it successfully, I will keep look at it and submit another PR to fix it. 

I try my best to make the test case arch-neutral as possible, but it seem not be good enough.
Could you please help to review it again? It's very close to code freezing date, I appreciate your comments for improving it~

**Comment by enjiamai** [comment]
@dcpleung excuse me, could you please help to review it when you have time, thanks!

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (dcpleung) [inline tests/kernel/interrupt/src/dynamic_isr.c:112]: This is about interrupt being triggered but should not be. So maybe something like `handler has run before interrupt trigger` which is more descriptive.

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (dcpleung) [inline tests/kernel/interrupt/src/dynamic_isr.c:118]: Something like `interrupt triggered but handler has not run`.

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (dcpleung) [inline tests/kernel/interrupt/src/dynamic_isr.c:123]: `parameter in handler is not correct`.

Might want to print out the argument too.

File: tests/kernel/interrupt/src/interrupt_util.h
Comment (dcpleung) [inline tests/kernel/interrupt/src/interrupt_util.h:104]: s/nubmer/number/

Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_util.h:104]: Got it fixed, thanks!

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/dynamic_isr.c:123]: Yes, got it fixed, thanks!

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/dynamic_isr.c:118]: Yes, got it fixed, thanks!

File: tests/kernel/interrupt/src/dynamic_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/dynamic_isr.c:112]: Yes, the words I used here is not adequate, just got it fix, thanks~

**Comment by enjiamai** [comment]
@dcpleung thank for your comments, I has already fixed all of them.

**Comment by enjiamai** [comment]
@LeiW000 Could you please help to review again, thanks!

**Review by dcpleung — APPROVED** [review]

**Review by LeiW000 — APPROVED** [review]

**Comment by enjiamai** [comment]
@stephanosio excuse me, could you please help to review it again when you have time? thanks!

File: tests/kernel/interrupt/src/interrupt_util.h
Comment (andyross) [inline tests/kernel/interrupt/src/interrupt_util.h:122]: This switch seems silly as the vectors are a simple function of the interrupt numbers.  Note that the other architectures you added don't try to interpret the numbers.   Why not:

```c
static inline void trigger_irq(int irq)
{
    int vector = irq + 32;
    __asm__ volatile("int %0" :: "i"(vector));
}
```

File: tests/kernel/interrupt/src/static_isr.c
Comment (andyross) [inline tests/kernel/interrupt/src/static_isr.c:188]: This doesn't look correct to me.  This is checking that "an address on the interrupt stack" is more than "main stack size" after "an address on this thread's stack".

What if the linker places the interrupt stack **lower** in memory than this stack?  What if the location of curr_stack_ptr is at the high address of the stack (i.e. it's a shallow call depth) and the int_0_stack_pointer is near the bottom, and the interrupt stack is immediately after?

The test here should be that the int_0_stack_pointer lies outside the main stack.  To do that you need to whitebox the thread struct and pull out the stack size and limit fields to check.

File: tests/kernel/interrupt/src/static_isr.c
Comment (andyross) [inline tests/kernel/interrupt/src/static_isr.c:80]: Too complicated.  You don't care about precision here, you just want "any address on the local stack".  So do it portably and take the address of a local variable:

```c
static inline void *get_stack_addr(uintptr_t *reg)
{
    int dummy;
    *reg = &dummy;
}
```

**Review by andyross — CHANGES_REQUESTED** [review]
Some nitpicks.  And unless I'm misunderstanding the setup there's a logic error in the stack bounds check.

File: tests/kernel/interrupt/src/interrupt_util.h
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_util.h:122]: Wow, it works. This stuck me for a long time, you just solve it in few seconds. thanks so much!

File: tests/kernel/interrupt/src/static_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/static_isr.c:188]: Got it, just let me think about it~

File: tests/kernel/interrupt/src/static_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/static_isr.c:80]: You are right, This is a good arch-neutral way to do that, learn a lot again, thanks!

File: tests/kernel/interrupt/src/static_isr.c
Comment (enjiamai) [reply tests/kernel/interrupt/src/static_isr.c]: @andyross I make some change of the checking logic, try to check if these two stack are overlap or not. It seems work well except qemu_cortex_a53 and native_posix board. And the stack information of qemu_cortex_a53 are :
 
start 0x400110c, size 0x1000, delta 0x0 (got it from &z_main_thread.stack_info) 

------ 0x400110c0 ----- main thread stack start -----
------ 0x4001107c ------ &dummy in interrupt context
------ 0x40010f4c ------ &dummy in thread context

so it's weird, seems like un-reasonable for me,  sorry, I was stuck here and still working on it.
I appreciate your comment if any, you always help me a lot~

**Comment by enjiamai** [comment]
Hi, @andyross I have fixed the logic error in the stack bounds check base on your comment. The way I applied here is to get the start and end address of the stack, then check if those two region overlap or not. Could you please help to review it again when you have time, thanks!

And I also enable the static interrupt test case qemu_xtensa, but only check one interrupt. Because there are only 2 SW interrupt can be used ( I confirm this from zephyr-sdk-0.11.4/xtensa/sample_controller/xtensa-zephyr-elf/include/arch/include/xtensa/config/core-isa.h ), and one has already been used by irq_offload.    

@andrewboie @andyross For the check failed in qemu_cortex_a53, native_posix and nrf52_bsim(posix), due to their stack pointer in interrupt context was not located in interrupt stack, I guess it might be a problem of qemu. I have to say sorry, as far I still can't figure out why and how to fix it at this moment, so I have to skip them now.

I am looking forward to getting any comment from both of you, thanks!

**Comment by LeiW000** [comment]
@enjiamai , what is status of the PR? Waitting for @andyross and @stephanosio to review?

**Comment by enjiamai** [comment]
@LeiW000 Yes, I think I have addressed all the comments they gave. But I guess they did not review it again due to I did not split them to a small commit. I will split them up within the week, Thanks!

**Comment by LeiW000** [comment]
@andyross, @stephanosio, could you please take look at the PR again? Enjia have addressed all your comments.

**Comment by enjiamai** [comment]
rework in another PR and close it.

**Pull request closed without merging**
